File: sdk/php/src/Client/Response.php
<?php
namespace Nmid\Client;

use Nmid\Constants;

// Client response decoder for PDT_S_RETURN_DATA.
// Mirrors node lib/client/response.js.
//
// body (PDT_S_RETURN_DATA):
//   HandleLen(4) | ParamsLen(4) | RetLen(4) | Handle | Params | Ret
class Response
{
    public int $dataType = 0;
    public string $data = '';
    public int $dataLen = 0;

    public string $handle = '';
    public int $handleLen = 0;

    public int $paramsLen = 0;
    public string $params = '';

    public string $ret = '';
    public int $retLen = 0;

    // Decode one complete packet (>= MIN_DATA_SIZE + dataLen bytes).
    public static function decodePack(string $packet): ?Response
    {
        $len = strlen($packet);
        if ($len < Constants::MIN_DATA_SIZE) {
            return null;
        }

        $header = unpack('N3', substr($packet, 0, Constants::MIN_DATA_SIZE));
        $dataType = $header[2];
        $dataLen = $header[3];

        if ($len < Constants::MIN_DATA_SIZE + $dataLen) {
            return null;
        }

        $resp = new self();
        $resp->dataType = $dataType;
        $resp->dataLen = $dataLen;
        $resp->data = substr($packet, Constants::MIN_DATA_SIZE, $dataLen);

        if ($dataType === Constants::PDT_S_RETURN_DATA) {
            $p = Constants::MIN_DATA_SIZE;
            $resp->handleLen = unpack('N', substr($packet, $p, 4))[1];
            $p += 4;
            $resp->paramsLen = unpack('N', substr($packet, $p, 4))[1];
            $p += 4;
            $resp->retLen = unpack('N', substr($packet, $p, 4))[1];
            $p += 4;
            $resp->handle = substr($packet, $p, $resp->handleLen);
            $p += $resp->handleLen;
            $resp->params = substr($packet, $p, $resp->paramsLen);
            $p += $resp->paramsLen;
            $resp->ret = substr($packet, $p, $resp->retLen);
        }

        return $resp;
    }

    public function getRet(): ?string
    {
        if ($this->dataType === Constants::PDT_S_RETURN_DATA) {
            return $this->ret;
        }
        return null;
    }
}

File: sdk/java/nmidsdk/client/src/main/java/com/nmidsdk/client/Response.java
package com.nmidsdk.client;

import com.nmidsdk.model.Constants;

import java.nio.charset.StandardCharsets;
import java.util.Arrays;

/**
 * Client response decoder.
 * Mirrors sdk/node/lib/client/response.js.
 *
 * For PDT_S_RETURN_DATA the body (read from the packet at MIN_DATA_SIZE) is:
 *   HandleLen(4) | ParamsLen(4) | RetLen(4) | Handle | Params | Ret
 */
public class Response {
    public int dataType;
    public byte[] data = new byte[0];
    public int dataLen;

    public String handle = "";
    public int handleLen;
    public int paramsLen;
    public byte[] params = new byte[0];
    public int retLen;
    public byte[] ret = new byte[0];

    public NmidError getResError() {
        switch (this.dataType) {
            case Constants.PDT_ERROR:
                return new NmidError("PDT_ERROR", "request error");
            case Constants.PDT_CANT_DO:
                return new NmidError("PDT_CANT_DO", "have no job do");
            case Constants.PDT_RATELIMIT:
                return new NmidError("PDT_RATELIMIT", "have ratelimit");
            default:
                return null;
        }
    }

    public byte[] getResResult() {
        if (this.dataType == Constants.PDT_S_RETURN_DATA) {
            return this.ret;
        }
        return null;
    }

    /**
     * Decode a single complete packet body into a Response.
     *
     * @param dataType the packet's data type (from the header)
     * @param dataLen  the packet's content length (from the header)
     * @param body     the body bytes (length == dataLen), WITHOUT the 12-byte header
     * @return the decoded Response, or null if the body is incomplete
     */
    public static Response decode(int dataType, int dataLen, byte[] body) {
        if (body == null || body.length < dataLen) {
            return null;
        }

        Response resp = new Response();
        resp.dataType = dataType;
        resp.dataLen = dataLen;
        resp.data = body;

        if (dataType == Constants.PDT_S_RETURN_DATA) {
            int start = 0;
            resp.handleLen = readIntBE(body, start);
            start += Constants.UINT32_SIZE;
            resp.paramsLen = readIntBE(body, start);
            start += Constants.UINT32_SIZE;
            resp.retLen = readIntBE(body, start);
            start += Constants.UINT32_SIZE;

            resp.handle = new String(body, start, resp.handleLen, StandardCharsets.UTF_8);
            start += resp.handleLen;

            resp.params = Arrays.copyOfRange(body, start, start + resp.paramsLen);
            start += resp.paramsLen;

            resp.ret = Arrays.copyOfRange(body, start, start + resp.retLen);
        }

        return resp;
    }

    static int readIntBE(byte[] b, int off) {
        return ((b[off] & 0xFF) << 24)
            | ((b[off + 1] & 0xFF) << 16)
            | ((b[off + 2] & 0xFF) << 8)
            | (b[off + 3] & 0xFF);
    }
}
